### PR TITLE
Make HandleError prevent hot-loops

### LINKED ISSUE #39816: [GC] controller manager gets error "unable to get REST mapping for kind" for ownerRefs to TPR and add-on APIs
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

TPR. GC. OwnerReference. REST mapping. 

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT.

**Kubernetes version** (use `kubectl version`):

master. v1.5.1

**Environment**:
- **Cloud provider or hardware configuration**: 
Repro in "hack/local-up-cluster.sh". But I have seen it in many different env.
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:
The controller manager is unresponsive and overloaded with error message:
```
E0112 18:26:11.690849   21374 garbagecollector.go:594] Error syncing item 
&garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"etcd.coreos.com/v1beta1", Kind:"Cluster", Name:"example-etcd-cluster", UID:"89603932-d8f4-11e6-94f3-42010af00002", Controller:(*bool)(0xc422850af8)}, Namespace:"default"}, dependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, dependents:map[*garbagecollector.node]struct {}{(*garbagecollector.node)(0xc42075a6c0):struct {}{}, (*garbagecollector.node)(0xc421a0abd0):struct {}{}, (*garbagecollector.node)(0xc421c074d0):struct {}{}}, owners:[]v1.OwnerReference(nil)}: 
unable to get REST mapping for kind: Cluster, version: etcd.coreos.com/v1beta1
```

This error just keeps bloating up without any backoff. This seems that REST mapping didn't recognize the kind.

**What you expected to happen**:

**How to reproduce it** (as minimally and precisely as possible):

Generic way:
- Create a TPR.
- Create a TPR object "A"
- Create pods that has OwnerReference pointing to "A".

Easy way:
- Use https://github.com/coreos/etcd-operator
- kubectl create -f example/deployment.yaml
- kubectl create -f example/example-etcd-cluster.yaml

**Anything else do we need to know**:

### PR DESCRIPTION
**Pull request by lavalamp**
Add an error "handler" that just sleeps for a bit if errors happen more
often than 500ms. Manually tested against #39816. This doesn't fix #39816 but it does keep it from crippling a cluster.

```release-note
Prevent hotloops on error conditions, which could fill up the disk faster than log rotation can free space.
```

### CODE REVIEW

**Comment by lavalamp** [comment]
I will make another PR with a test. This is the minimum fix and it should be easy to cherrypick.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go [hunk 1]
```diff
@@ -101,6 +111,26 @@ func logError(err error) {
 	glog.ErrorDepth(2, err)
 }
 
+type rudimentaryErrorBackoff struct {
+	// TODO(lavalamp): use the clock for testability. Need to move that
+	// package for that to be accessible here.
+	lastErrorTime time.Time
+	minPeriod     time.Duration
+	lock          sync.Mutex
```
Comment (mikedanese) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: nit, lock should be the field above the thing it locks and should be named lasteErrorTimeLock or something.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go [hunk 2]
```diff
@@ -101,6 +111,26 @@ func logError(err error) {
...
+}
+
+// OnError will block if it is called more often than the embedded period time.
+// This will prevent overly tight hot error loops.
+func (r *rudimentaryErrorBackoff) OnError(error) {
```
Comment (mikedanese) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: Now we aren't even going to log the error? Is this responsible for more error logs then just the one we were seeing in logspam?

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go (hunk 1 above)
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: ok, will fix here and in the cherrypick branch

Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: cherrypick branch is fixed.

**Comment by lavalamp** [comment]
I want to make the fix in the release branch here before this merges.

**Comment by lavalamp** [comment]
(but I'm building and testing the release branch first)

**Comment by lavalamp** [comment]
OK, this should be mergable now.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go [hunk 3]
```diff
@@ -79,7 +81,15 @@ func getCallers(r interface{}) string {
 
 // ErrorHandlers is a list of functions which will be invoked when an unreturnable
 // error occurs.
-var ErrorHandlers = []func(error){logError}
+// TODO(lavalamp): for testability, this and the below HandleError function
+// should be packaged up into a testable and reusable object.
+var ErrorHandlers = []func(error){
+	logError,
+	(&rudimentaryErrorBackoff{
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: I'm not a fan of this solution.  Other controllers use a backoff requeue mechanism in the work queue instead.  A global backoff handler for the entire process doesn't seem like a reasonable solution.

**Comment by deads2k** [comment]
Can we consider a solution that is targeted to the area that had problems instead of globally affecting everyone?

@ncdc @liggitt @derekwaynecarr @kargakis You guys are all in the controller code too.

**Comment by deads2k** [comment]
Added a tag to resolve questions about alternative solutions before this merges and affects all callers.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go (hunk 3 above)
Comment (kargakis) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: Ideally, I would like to see us formalizing error handling in the controller framework. Error handling differs between controllers and this is indeed a change that affects all controllers which is not good. Most of them already use a backoff mechanism as @deads2k mentioned.

**Comment by ncdc** [comment]
I'll second what @deads2k said. Let's focus on the code that's hot-looping and formalize our error handling functionality in the controller framework.

**Comment by deads2k** [comment]
> I'll second what @deads2k said. Let's focus on the code that's hot-looping and formalize our error handling functionality in the controller framework.

`.AddRateLimited(` is pretty common.  apiservice, certificate, daemon, deployment, disruption, endpoints, jobs, namespaces, replicaset, quota, service accounts, tokens, statefulset.

**Comment by smarterclayton** [comment]
So there's two different arguments going on here:

1. The process has a spout that is unbounded (handle error) which is intended to be used for handling "shrug" scenarios (literally, we fail, and it's better than eating the error) - we want all code to use that spout, but the spout *must* be rate limited
2. Controllers already have a mechanism for this, but it's not being used

David is concerned that 2 should be fixed at the same time as 1, and that 1 makes 2 not work as well (because it allows other code that is not controllers to break controllers).  I agree with this, but we'd need to police the mechanism for 1.

So:

1. We must rate limit util.HandleError to non-system destructive levels (arguably this should be per handler, not global, but whatever)
2. We need a way to allow proper rate limited things like controllers to bypass 1 safely
3. We need to prevent other people in the system from abusing 2 to bypass 1.  That's effectively something like import boss.

**Comment by smarterclayton** [comment]
Doesn't glog have a rate limiter?  Why wouldn't we also set that?

**Comment by deads2k** [comment]
> Controllers already have a mechanism for this, but it's not being used

It's be used by the majority, it's not being used by GC.  This effectively breaks the backoff handling of most controllers in a way that unnecessarily blocks execution of that controller (AddRateLimited doesn't) and the majority of callers from the controller packages don't want a delay like this.

Further, if you're building something to just stop a hotloop on an unconsidered error (again, not the case in the majority of controllers), you can sleep for a very short period (10s of milliseconds) and you'd just do it unconditionally since the purpose is to just avoid ddos-ing yourself.  However, since its the opposite of what the majority of controllers want, all the existing controllers with proper handling need to be updated to *not* use this new (or severely changed) method.

Ending up in this place means that instead of fixing rating limiting for the GC controller (clients have rate limiters) and instead of fixing the GC controller to AddRateLimited (libraries already exist), means that we've taken the biggest hammer we have and applied it across multiple processes and allowed errors in on go routine to negatively rate limit already rate limited controllers for one bad apple.

**Comment by smarterclayton** [comment]
> However, since its the opposite of what the majority of controllers want, all the existing controllers with proper handling need to be updated to not use this new (or severely changed) method.

Yes, that's 2, and we need 3 to prevent it from being abused.

> Ending up in this place means that instead of fixing rating limiting for the GC controller (clients have rate limiters) and instead of fixing the GC controller to AddRateLimited (libraries already exist), means that we've taken the biggest hammer we have and applied it across multiple processes and allowed errors in on go routine to negatively rate limit already rate limited controllers for one bad apple.

I had the expectation that 

1. GC gets fixed
2. We add the new method and use it in controllers
3. We fix controllers that are sending too much to handle error to more accurately ignore common errors
4. We reduce the delay imposed here to something much smaller.

I would expect us to fix that in 1.6.

**Comment by deads2k** [comment]
https://github.com/kubernetes/kubernetes/pull/38679 switches GC to a ratelimited work queue for 1.6 and simple wait could be added here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go#L594 as a minimal touch in 1.5 instead of disrupting most of the other controllers in the 1.5 stream.

**Comment by lavalamp** [comment]
I did this because I never want to have an afternoon destroyed by someone logging every 20 microseconds again, and this seemed like the most general place with the fewest side effects. I don't care if we make it 50ms instead of 500. Should we really be `HandleError`ing that frequently? I think that's a problem even if it's separate components.

**Comment by deads2k** [comment]
We spoke in slack and decided that a 1ms delay would protect infrastructure with minimal impact to callers like controllers.  Once its updated, I'm ok with sleeping here.

**Comment by mikedanese** [comment]
@lavalamp can you cherrypick into 1.4 while you are at it?

**Comment by lavalamp** [comment]
We had a discussion on slack and it seems 1ms is the number that makes everyone able to live with a global limit like this. I will update this PR and send an adjustment to the 1.5 branch.

**Comment by lavalamp** [comment]
OK, number adjusted, in a second commit so it'll be easier to cherrypick.

**Comment by lavalamp** [comment]
@grodrigues3 @apelisse The bot seems confused about the LGTM ordering here? Or is there some other reason why this is stuck?

**Comment by apelisse** [comment]
Agreed, there is a confusion. Thanks Lavalamp, somebody noticed this bug before but we couldn't figure out what was going wrong. I think I somewhat understand now.

**Pull request merged**